upkie/cpp/actuation/BulletInterface.cpp

```diff
@@ -344,7 +346,27 @@ void BulletInterface::read_contacts() {
         contact_info.m_numContactPoints;
   }
 }
-
+int BulletInterface::environment_collision() {
+  b3ContactInformation contact_info;
+  b3RobotSimulatorGetContactPointsArgs contact_args;
+  const int nb_links = bullet_.getNumJoints(robot_);
+  for (const auto& key_child : body_names) {
+    const auto& env_id = key_child.second;
+    for (int link_id = 0; link_id < nb_links; ++link_id) {
+      if (get_link_index("left_wheel_tire") != link_id &&
+          get_link_index("right_wheel_tire") != link_id) {
+        contact_args.m_bodyUniqueIdA = robot_;
+        contact_args.m_bodyUniqueIdB = env_id;
+        contact_args.m_linkIndexA = link_id;
+        bullet_.getContactPoints(contact_args, &contact_info);
```

Comment:
Hmm, that's a quadratic number of calls to `getContactPoints`. Does it impact `spine.clock.slack` already on your machine?

I guess it shouldn't impact us so much now, but this may not scale well. We could make this a linear (in the number of Upkie bodies, not env bodies) by using the contact point cache, similarly to what is done for instance in [this Bullet example](https://github.com/bulletphysics/bullet3/blob/e9c461b0ace140d5c73972760781d94b7b5eee53/examples/Evolution/NN3DWalkers.cpp#L753):

```c++
	int numManifolds = m_dynamicsWorld->getDispatcher()->getNumManifolds();
	for (int i = 0; i < numManifolds; i++)
	{
		btPersistentManifold* contactManifold = m_dynamicsWorld->getDispatcher()->getManifoldByIndexInternal(i);
		const btCollisionObject* obA = contactManifold->getBody0();
		const btCollisionObject* obB = contactManifold->getBody1();

		if (obA->getUserPointer() != GROUND_ID && obB->getUserPointer() != GROUND_ID)
		{
			int numContacts = contactManifold->getNumContacts();
```

Reply:
Resolving this conversation in favor of the alternative suggested below.